synapse/storage/util/id_generators.py

```diff
@@ -446,23 +446,31 @@ def _load_current_ids(
 
             rows: List[Tuple[str, int]] = []
             for table, instance_column, id_column in tables:
+                logger.info(
+                    "table=%s instance_column=%s id_column=%s",
+                    table,
+                    instance_column,
+                    id_column,
+                )
                 sql = """
                     SELECT %(instance)s, %(id)s FROM %(table)s
                     WHERE ? %(cmp)s %(id)s
+                    /* Sort so that we handle rows in order for each instance. */
+                    ORDER BY %(id)s ASC
                 """ % {
                     "id": id_column,
                     "table": table,
                     "instance": instance_column,
                     "cmp": "<=" if self._positive else ">=",
                 }
+                logger.info(
+                    "sql=%s args=%s", sql, (min_stream_id * self._return_factor,)
+                )
                 cur.execute(sql, (min_stream_id * self._return_factor,))
 
                 # Cast safety: this corresponds to the types returned by the query above.
                 rows.extend(cast(Iterable[Tuple[str, int]], cur))
 
-            # Sort so that we handle rows in order for each instance.
```

Comment:
What is the correct order that we wanted?

The previous code sorted it by `instance_name` because that was the first column selected.

Reply:
(requesting review for this comment)

Reply:
My best guest is the line:

> self._current_positions[instance] = stream_id

Where if we *didn't* ensure that we handled the rows in ascending stream ID order (per-instance) then we could end up setting the current position to a lower stream ID than we ought to. 

I think its enough to do `rows.sort(key=lambda _, stream_id: stream_id)` or similar.

Reply:
Thanks for the context @erikjohnston!

I don't know of the symptoms this would manifest as or underlying problems this causes but I guess this fixes some other bugs with incorrect `stream_ids` (all the places we use `MultiWriterIdGenerator`).

Probably solving some duplicate primary key database errors at the very least. And maybe larger bugs like `to_device` messages not sending.